psh/wget/wget.c

```diff
@@ -0,0 +1,489 @@
+
+	tmp = strchr(bare, '/');
+	common.path = tmp + 1;
+	common.host = strndup(bare, tmp - bare);
+
+	return 0;
+}
+
+
+static uint32_t *wget_aiToAddr(struct addrinfo *ai)
+{
+	return (uint32_t *)&((struct sockaddr_in *)ai->ai_addr)->sin_addr;
+}
+
+
+static int wget_connect(const char *host)
+{
+	struct addrinfo *res;
+	struct addrinfo hints = { 0 };
+	char hostaddr[INET_ADDRSTRLEN];
+	int fd = -1;
+
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+
+	printf("Resolving %s... ", host);
+	fflush(stdout);
+	if (getaddrinfo(host, "80", &hints, &res) != 0) {
+		printf("Failed\n");
+		return -1;
+	}
+
+	if (inet_ntop(res->ai_family, wget_aiToAddr(res), hostaddr, sizeof(hostaddr)) == NULL) {
+		fprintf(stderr, "nslookup: Error while converting address!\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("%s\n", hostaddr);
+
+	printf("Connecting to %s|%s|:80... ", host, hostaddr);
+	fflush(stdout);
+
+	fd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+	if (fd < 0) {
+		printf("Failed\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+
+	if (connect(fd, res->ai_addr, res->ai_addrlen) < 0) {
+		close(fd);
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("Connected\n");
+
+	freeaddrinfo(res);
+
+	return fd;
+}
+
+
+static int wget_request(void)
+{
+	char buf[256];
+	int bytes = 0, i = 0, ret;
+
+	bytes += sprintf(buf, "GET /%s HTTP/1.1\r\n", common.path);
+	bytes += sprintf(buf + bytes, "Host: %s\r\nUser-Agent: Wget\r\n", common.host);
+	bytes += sprintf(buf + bytes, "\r\n");
+
+	while (i < bytes) {
+		ret = write(common.socket, buf + i, bytes - i);
+		if (ret < 0) {
+			return -1;
+		}
+
+		i += ret;
+	}
+
+	if (common.debug) {
+		printf("--------Request--------\n%*s\n", bytes, buf);
+	}
+
+	return 0;
+}
+
+
+static int wget_file(size_t len)
+{
+	int ret;
+	size_t left = len;
+	int i = 0;
+
+	/* Save, what is left in the buffer */
+	if (common.readptr < common.buf + common.len) {
+		ret = fwrite(common.readptr, 1, common.buf + common.len - common.readptr, common.outfile);
+		if (ret < 0) {
+			fprintf(stderr, "wget: Write failed\n");
+			return -1;
+		}
+		left -= ret;
+	}
+
+	while (left > 0) {
+		ret = read(common.socket, common.buf, min(sizeof(common.buf), left));
+		if (ret < 0) {
+			fprintf(stderr, "wget: read from socket failed!\n");
+			return -1;
+		}
+
+		left -= ret;
+
+		ret = fwrite(common.buf, 1, ret, common.outfile);
+		if (ret < 0) {
+			fprintf(stderr, "wget: Write failed\n");
+			return -1;
+		}
+		if (i % 8 == 0) {
+			printf("\rWritten: %8d/%d", len - left, len);
+			fflush(stdout);
+		}
+		i = (i + 1) % 8;
+	}
+	printf("\rWritten: %8d/%d", len - left, len);
+	fflush(stdout);
+	putchar('\n');
+
+	return len - left;
+}
+
+
+static int wget_buf(void)
+{
+	char *dst;
+	int cnt;
+	int ret;
+
+	dst = common.buf;
+
+	cnt = 0;
+	/* Move not consumed part of the buffer */
+	while (common.readptr < common.buf + common.len) {
+		*(dst++) = *(common.readptr++);
+		cnt++;
+	}
```

Comment:
Not exactly. We are moving bytes, which are not consumed yet, to handle the case, when the buffer is full and we receive an incomplete header -  a string without "\r\n" at the end. Src and dst of this operation point to the same buffer, so they could potentially overlap, so using `memcpy()` would cause an undefined behavior. It look like the best solution here would be to use `memmove()`.